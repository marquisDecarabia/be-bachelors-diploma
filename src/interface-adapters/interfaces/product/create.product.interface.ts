export interface CreateProduct {
  name: string;
  price: number;
  currency: string;
}
