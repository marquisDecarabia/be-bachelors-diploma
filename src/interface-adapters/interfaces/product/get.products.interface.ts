export interface GetProducts {
  limit: number;
  page: number;
}
