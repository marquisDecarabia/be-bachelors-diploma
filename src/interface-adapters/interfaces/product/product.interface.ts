import { ModelBase } from '@libs/ddd/interface-adapters/interfaces/model.base.interface';

export interface Product extends ModelBase {
  name: string;
  price: number;
  status: string;
  currency: string;
}
