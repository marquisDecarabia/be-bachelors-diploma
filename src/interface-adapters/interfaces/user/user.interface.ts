import { ModelBase } from '@libs/ddd/interface-adapters/interfaces/model.base.interface';

export interface User extends ModelBase {
  name: string;
}
