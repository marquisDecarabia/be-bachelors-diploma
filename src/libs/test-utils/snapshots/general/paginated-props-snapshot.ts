export const paginatedPropsSnapshot = {
  data: expect.any(Array),
  count: expect.any(Number),
  limit: expect.any(Number),
  page: expect.any(Number),
};
