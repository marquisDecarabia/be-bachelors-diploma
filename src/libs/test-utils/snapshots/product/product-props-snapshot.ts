export const productPropsSnapshot = {
  id: expect.any(String),
  createdAt: expect.any(String),
  updatedAt: expect.any(String),
  name: expect.any(String),
  price: expect.any(Number),
  status: expect.any(String),
  currency: expect.any(String),
};

export const productPropsSnapshotWithParams = (
  name: string,
  price: number,
  currency: string,
  status: string,
) => {
  return {
    ...productPropsSnapshot,
    name: name,
    price: price,
    currency: currency,
    status: status,
  };
};
