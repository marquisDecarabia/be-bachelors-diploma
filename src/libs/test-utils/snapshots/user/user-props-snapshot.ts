export const userPropsSnapshot = {
  id: expect.any(String),
  createdAt: expect.any(String),
  updatedAt: expect.any(String),
  name: expect.any(String),
};
