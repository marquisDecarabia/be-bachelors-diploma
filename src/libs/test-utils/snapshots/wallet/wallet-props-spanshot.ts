export const walletPropsSnapshot = {
  id: expect.any(String),
  createdAt: expect.any(String),
  updatedAt: expect.any(String),
  userId: expect.any(String),
  balance: expect.any(Number),
  currency: expect.any(String),
};

export const walletPropsSnapshotWithParams = (
  userId: string,
  balanceNum: number,
  currency: string,
) => {
  return {
    ...walletPropsSnapshot,
    userId: userId,
    balance: balanceNum,
    currency: currency,
  };
};
