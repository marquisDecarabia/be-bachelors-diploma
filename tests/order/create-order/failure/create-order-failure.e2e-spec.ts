import { defineFeature, loadFeature } from 'jest-cucumber';
import { getTestServer, TestServer } from '../../../jestSetupAfterEnv';
import * as request from 'supertest';
import { Id } from '@libs/ddd/interface-adapters/interfaces/id.interface';
import { Product } from '@src/interface-adapters/interfaces/product/product.interface';
import { userPropsSnapshot } from '@libs/test-utils/snapshots/user/user-props-snapshot';
import {
  errorResponseSnapshot,
  errorResponseSnapshotWParams,
} from '@libs/test-utils/snapshots/general/error-response-snapshot';
import { walletPropsSnapshotWithParams } from '@libs/test-utils/snapshots/wallet/wallet-props-spanshot';
import { paginatedPropsSnapshot } from '@libs/test-utils/snapshots/general/paginated-props-snapshot';
import { productPropsSnapshot } from '@libs/test-utils/snapshots/product/product-props-snapshot';

const feature = loadFeature(
  'tests/order/create-order/failure/create-order-failure.feature',
);

defineFeature(feature, (test) => {
  let testServer: TestServer;
  let httpServer: request.SuperTest<request.Test>;

  beforeAll(() => {
    testServer = getTestServer();
    httpServer = request(testServer.serverApplication.getHttpServer());
  });

  test('Creating an order (failure)', ({ given, and, when, then }) => {
    let userId: Id;
    let products: Product[];
    let product: Product;
    let orderResponse;

    given(/^that my user ID is "(.*)"$/, async (id: string) => {
      const result = await httpServer.get(`/v1/users/${id}`).send().expect(200);
      expect(result.body).toMatchSnapshot(userPropsSnapshot);
      userId = { id };
    });

    and(
      /^my wallet balance is "(.*)" "(.*)"$/,
      async (balance: string, currency: string) => {
        const numBalance = Number(balance);
        const res = await httpServer
          .get(`/v1/wallets/user/${userId.id}`)
          .send()
          .expect(200);
        expect(res.body).toMatchSnapshot(
          walletPropsSnapshotWithParams(userId.id, numBalance, currency),
        );
      },
    );

    when('i request products list', async () => {
      const res = await httpServer
        .get('/v1/products?page=1&limit=20')
        .send()
        .expect(200);
      expect(res.body).toMatchSnapshot(paginatedPropsSnapshot);
      res.body.data.forEach((product) => {
        expect(product).toMatchSnapshot(productPropsSnapshot);
      });
      products = res.body.data;
    });

    then(
      /^I see product with name is "(.*)" and price is "(.*)" "(.*)" and status is "(.*)"$/,
      (productName, price, currency, status) => {
        const numPrice = Number(price);
        product = products.find((item) => {
          return (
            item.name === productName &&
            item.price === numPrice &&
            item.currency === currency &&
            item.status === status
          );
        });
        expect(product).toMatchSnapshot(productPropsSnapshot);
      },
    );

    when('I send a request to create an order with this product', async () => {
      orderResponse = await httpServer.post('/v1/orders').send({
        userId: userId.id,
        productId: product.id,
      });
    });

    then('I receive error response', () => {
      expect(orderResponse.body).toMatchSnapshot(errorResponseSnapshot);
    });

    and(
      /^response message is "(.*)" and code is "(.*)" and error is "(.*)"$/,
      (message: string, code: string, error: string) => {
        const numCode = Number(code);
        expect(orderResponse.body).toMatchSnapshot(
          errorResponseSnapshotWParams(numCode, message, error),
        );
      },
    );
  });
});
