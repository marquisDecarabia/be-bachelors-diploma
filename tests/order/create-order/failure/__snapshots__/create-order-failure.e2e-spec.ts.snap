// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create an order (failure) Creating an order (failure) 1`] = `
Object {
  "createdAt": Any<String>,
  "id": Any<String>,
  "name": Any<String>,
  "updatedAt": Any<String>,
}
`;

exports[`Create an order (failure) Creating an order (failure) 2`] = `
Object {
  "balance": 119.91,
  "createdAt": Any<String>,
  "currency": "USD",
  "id": Any<String>,
  "updatedAt": Any<String>,
  "userId": "11299efa-a2ee-4f65-857e-2fff02d2446e",
}
`;

exports[`Create an order (failure) Creating an order (failure) 3`] = `
Object {
  "count": Any<Number>,
  "data": Any<Array>,
  "limit": Any<Number>,
  "page": Any<Number>,
}
`;

exports[`Create an order (failure) Creating an order (failure) 4`] = `
Object {
  "createdAt": Any<String>,
  "currency": Any<String>,
  "id": Any<String>,
  "name": Any<String>,
  "price": Any<Number>,
  "status": Any<String>,
  "updatedAt": Any<String>,
}
`;

exports[`Create an order (failure) Creating an order (failure) 5`] = `
Object {
  "createdAt": Any<String>,
  "currency": Any<String>,
  "id": Any<String>,
  "name": Any<String>,
  "price": Any<Number>,
  "status": Any<String>,
  "updatedAt": Any<String>,
}
`;

exports[`Create an order (failure) Creating an order (failure) 6`] = `
Object {
  "createdAt": Any<String>,
  "currency": Any<String>,
  "id": Any<String>,
  "name": Any<String>,
  "price": Any<Number>,
  "status": Any<String>,
  "updatedAt": Any<String>,
}
`;

exports[`Create an order (failure) Creating an order (failure) 7`] = `
Object {
  "createdAt": Any<String>,
  "currency": Any<String>,
  "id": Any<String>,
  "name": Any<String>,
  "price": Any<Number>,
  "status": Any<String>,
  "updatedAt": Any<String>,
}
`;

exports[`Create an order (failure) Creating an order (failure) 8`] = `
Object {
  "createdAt": Any<String>,
  "currency": Any<String>,
  "id": Any<String>,
  "name": Any<String>,
  "price": Any<Number>,
  "status": Any<String>,
  "updatedAt": Any<String>,
}
`;

exports[`Create an order (failure) Creating an order (failure) 9`] = `
Object {
  "createdAt": Any<String>,
  "currency": Any<String>,
  "id": Any<String>,
  "name": Any<String>,
  "price": Any<Number>,
  "status": Any<String>,
  "updatedAt": Any<String>,
}
`;

exports[`Create an order (failure) Creating an order (failure) 10`] = `
Object {
  "createdAt": Any<String>,
  "currency": Any<String>,
  "id": Any<String>,
  "name": Any<String>,
  "price": Any<Number>,
  "status": Any<String>,
  "updatedAt": Any<String>,
}
`;

exports[`Create an order (failure) Creating an order (failure) 11`] = `
Object {
  "createdAt": Any<String>,
  "currency": Any<String>,
  "id": Any<String>,
  "name": Any<String>,
  "price": Any<Number>,
  "status": Any<String>,
  "updatedAt": Any<String>,
}
`;

exports[`Create an order (failure) Creating an order (failure) 12`] = `
Object {
  "createdAt": Any<String>,
  "currency": Any<String>,
  "id": Any<String>,
  "name": Any<String>,
  "price": Any<Number>,
  "status": Any<String>,
  "updatedAt": Any<String>,
}
`;

exports[`Create an order (failure) Creating an order (failure) 13`] = `
Object {
  "error": Any<String>,
  "message": Any<String>,
  "statusCode": Any<Number>,
}
`;

exports[`Create an order (failure) Creating an order (failure) 14`] = `
Object {
  "error": "Conflict",
  "message": "Product is unavailable",
  "statusCode": 409,
}
`;

exports[`Create an order (failure) Creating an order (failure) 15`] = `
Object {
  "createdAt": Any<String>,
  "id": Any<String>,
  "name": Any<String>,
  "updatedAt": Any<String>,
}
`;

exports[`Create an order (failure) Creating an order (failure) 16`] = `
Object {
  "balance": 119.91,
  "createdAt": Any<String>,
  "currency": "USD",
  "id": Any<String>,
  "updatedAt": Any<String>,
  "userId": "11299efa-a2ee-4f65-857e-2fff02d2446e",
}
`;

exports[`Create an order (failure) Creating an order (failure) 17`] = `
Object {
  "count": Any<Number>,
  "data": Any<Array>,
  "limit": Any<Number>,
  "page": Any<Number>,
}
`;

exports[`Create an order (failure) Creating an order (failure) 18`] = `
Object {
  "createdAt": Any<String>,
  "currency": Any<String>,
  "id": Any<String>,
  "name": Any<String>,
  "price": Any<Number>,
  "status": Any<String>,
  "updatedAt": Any<String>,
}
`;

exports[`Create an order (failure) Creating an order (failure) 19`] = `
Object {
  "createdAt": Any<String>,
  "currency": Any<String>,
  "id": Any<String>,
  "name": Any<String>,
  "price": Any<Number>,
  "status": Any<String>,
  "updatedAt": Any<String>,
}
`;

exports[`Create an order (failure) Creating an order (failure) 20`] = `
Object {
  "createdAt": Any<String>,
  "currency": Any<String>,
  "id": Any<String>,
  "name": Any<String>,
  "price": Any<Number>,
  "status": Any<String>,
  "updatedAt": Any<String>,
}
`;

exports[`Create an order (failure) Creating an order (failure) 21`] = `
Object {
  "createdAt": Any<String>,
  "currency": Any<String>,
  "id": Any<String>,
  "name": Any<String>,
  "price": Any<Number>,
  "status": Any<String>,
  "updatedAt": Any<String>,
}
`;

exports[`Create an order (failure) Creating an order (failure) 22`] = `
Object {
  "createdAt": Any<String>,
  "currency": Any<String>,
  "id": Any<String>,
  "name": Any<String>,
  "price": Any<Number>,
  "status": Any<String>,
  "updatedAt": Any<String>,
}
`;

exports[`Create an order (failure) Creating an order (failure) 23`] = `
Object {
  "createdAt": Any<String>,
  "currency": Any<String>,
  "id": Any<String>,
  "name": Any<String>,
  "price": Any<Number>,
  "status": Any<String>,
  "updatedAt": Any<String>,
}
`;

exports[`Create an order (failure) Creating an order (failure) 24`] = `
Object {
  "createdAt": Any<String>,
  "currency": Any<String>,
  "id": Any<String>,
  "name": Any<String>,
  "price": Any<Number>,
  "status": Any<String>,
  "updatedAt": Any<String>,
}
`;

exports[`Create an order (failure) Creating an order (failure) 25`] = `
Object {
  "createdAt": Any<String>,
  "currency": Any<String>,
  "id": Any<String>,
  "name": Any<String>,
  "price": Any<Number>,
  "status": Any<String>,
  "updatedAt": Any<String>,
}
`;

exports[`Create an order (failure) Creating an order (failure) 26`] = `
Object {
  "createdAt": Any<String>,
  "currency": Any<String>,
  "id": Any<String>,
  "name": Any<String>,
  "price": Any<Number>,
  "status": Any<String>,
  "updatedAt": Any<String>,
}
`;

exports[`Create an order (failure) Creating an order (failure) 27`] = `
Object {
  "error": Any<String>,
  "message": Any<String>,
  "statusCode": Any<Number>,
}
`;

exports[`Create an order (failure) Creating an order (failure) 28`] = `
Object {
  "error": "Bad Request",
  "message": "Not enough funds to buy product",
  "statusCode": 400,
}
`;
